php-frontend/src/main/java/org/sonar/php/parser/TreeFactory.java

```diff
@@ -0,0 +1,1100 @@
+      elements.add(new IdentifierTreeImpl(namespaceToken));
+      separators.add(separator);
+    }
+
+    if (listOptional.isPresent()) {
+      for (Tuple<InternalSyntaxToken, InternalSyntaxToken> tuple : listOptional.get()) {
+        elements.add(new IdentifierTreeImpl(tuple.first()));
+        separators.add(tuple.second());
+      }
+    }
+
+    return new NamespaceNameTreeImpl(absoluteSeparator, new SeparatedList<>(elements.build(), separators.build()), new IdentifierTreeImpl(name));
+
+  }
+
+  public CatchBlockTree catchBlock(InternalSyntaxToken catchToken, InternalSyntaxToken lParenthesis, NamespaceNameTree exceptionType, InternalSyntaxToken variable, InternalSyntaxToken rParenthsis, BlockTree block) {
+    return new CatchBlockTreeImpl(
+        catchToken,
+        lParenthesis,
+        exceptionType,
+        new VariableIdentifierTreeImpl(new IdentifierTreeImpl(variable)),
+        rParenthsis,
+        block
+    );
+  }
+
+  public EmptyStatementTree emptyStatement(InternalSyntaxToken semicolonToken) {
+    return new EmptyStatementImpl(semicolonToken);
+  }
+
+  public ThrowStatementTree throwStatement(InternalSyntaxToken throwToken, ExpressionTree expression, InternalSyntaxToken eosToken) {
+    return new ThrowStatementTreeImpl(throwToken, expression, eosToken);
+  }
+
+  public ForEachStatementTree forEachStatement(ForEachStatementHeader header, StatementTree statement) {
+    return new ForEachStatementTreeImpl(header, statement);
+  }
+
+  public ForEachStatementTree forEachStatementAlternative(
+      ForEachStatementHeader header,
+      InternalSyntaxToken colonToken, Optional<List<StatementTree>> statements, InternalSyntaxToken endForEachToken, InternalSyntaxToken eosToken
+  ) {
+    return new ForEachStatementTreeImpl(header, colonToken, optionalList(statements), endForEachToken, eosToken);
+  }
+
+  public ForEachStatementHeader forEachStatementHeader(
+      InternalSyntaxToken forEachToken, InternalSyntaxToken openParenthesisToken,
+      ExpressionTree expression, InternalSyntaxToken asToken, Optional<Tuple<ExpressionTree, InternalSyntaxToken>> optionalKey, ExpressionTree value,
+      InternalSyntaxToken closeParenthesisToken
+  ) {
+    return new ForEachStatementHeader(
+        forEachToken, openParenthesisToken,
+        expression, asToken, getForEachKey(optionalKey), getForEachArrow(optionalKey), value,
+        closeParenthesisToken
+    );
+  }
+
+  @Nullable
+  private ExpressionTree getForEachKey(Optional<Tuple<ExpressionTree, InternalSyntaxToken>> optionalKey) {
+    if (optionalKey.isPresent()) {
+      return optionalKey.get().first();
+    } else {
+      return null;
+    }
+  }
+
+  @Nullable
+  private InternalSyntaxToken getForEachArrow(Optional<Tuple<ExpressionTree, InternalSyntaxToken>> optionalKey) {
+    if (optionalKey.isPresent()) {
+      return optionalKey.get().second();
+    } else {
+      return null;
+    }
+  }
+
+  public ForStatementHeader forStatementHeader(
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Method has 8 parameters, which is greater than 7 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS00107)